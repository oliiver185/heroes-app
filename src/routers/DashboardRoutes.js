import React from 'react'
import { Redirect, Route, Switch } from 'react-router-dom'
import { DcScreen } from '../components/dc/DcScreen'
import { HeroScreen } from '../components/heros/HeroScreen'
import { MarvelScreen } from '../components/marvel/MarvelScreen'
import { SearchScreen } from '../components/search/SearchScreen'
import { Navbar } from '../components/ui/Navbar'

export const DashboardRoutes = () => {
    return (
        <>
           

          <Navbar />  
           <div className="marvel-background">

          <div className="container p-3 ">

              <Switch>
                  <Route exact path="/marvel" component={MarvelScreen}/>
                  <Route exact path="/hero/:heroId" component={HeroScreen}/>
                  {/* /:heroId, lo que significa que podemos usar el useParamsHook más adelante para obtener el parámetro de ruta. */}
                  <Route exact path="/dc" component={DcScreen}/>
                  <Route exact path="/search" component={SearchScreen}/>

                  <Redirect to="/marvel"/>
              </Switch>
          </div>
           </div>
        </>
    )
}

// the only difference between the first router is that this does not include <Router><Router>