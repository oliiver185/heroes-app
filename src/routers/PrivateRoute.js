import React from 'react';
import { Redirect, Route } from 'react-router-dom';
import PropTypes from 'prop-types';

//aqui cierro y abro llaves porque voy a recibir
//varios componenetes en las properties
export const PrivateRoute = ({
    isAuthenticated,
    component: Component,
    ...rest //son todos los demas argumentos como e
    //el exact path etc necesito recuperarlos con el rest 
    //y puedo pasarlo al componente
}) => {


    localStorage.setItem('lastPath', rest.location.pathname);
    return (

        <Route
            {...rest}
            component={(props) => ( //estamos recibiendo las pros history location etc

                (isAuthenticated)
                    ?
                    (<Component {...props} />)
                    :
                    (<Redirect to="/login" />)
            )
            }

        />

    )
}

// ponemos parentesis porque solo vamos a renderizar alog 


PrivateRoute.propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    component: PropTypes.func.isRequired
}