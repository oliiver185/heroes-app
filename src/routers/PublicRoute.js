import React from 'react';
import { Redirect, Route } from 'react-router-dom';
import PropTypes from 'prop-types';

//aqui cierro y abro llaves porque voy a recibir
//varios componenetes en las properties
export const PublicRoute = ({
    isAuthenticated,
    component: Component,
    ...rest //son todos los demas argumentos como e
    //el exact path etc necesito recuperarlos con el rest 
    //y puedo pasarlo al componente
}) => {

    return (

        <Route
            {...rest}
            component={(props) => ( //estamos recibiendo las pros history location etc

                (!isAuthenticated)
                    ?
                    (<Component {...props} />)
                    :
                    (<Redirect to="/" />)
            )
            }

        />

    )
}

// ponemos parentesis porque solo vamos a renderizar alog 


PublicRoute.propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    component: PropTypes.func.isRequired
}