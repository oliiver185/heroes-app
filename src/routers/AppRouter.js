import React, { useContext } from "react";
import {
  HashRouter as Router,
  Switch

} from "react-router-dom";
import { PrivateRoute } from "./PrivateRoute";
import { LoginScreen } from "../components/login/LoginScreen";
import { DashboardRoutes } from "./DashboardRoutes";
// import { authReducer } from "../auth/authReducer";
import { AuthContext } from "../auth/AuthContext";
import { PublicRoute } from "./PublicRoute";


export const AppRouter = () => {
   
  
  const {user} = useContext(AuthContext);
  // console.log(user);
  
  return (
        <Router>
        <div>

  
         
          <Switch>
        
           <PublicRoute
            exact
             path="/login" 
             isAuthenticated={user.logged}
             component={LoginScreen}/> 


           
           {/* //vamos a cambiar route por private route */}
           <PrivateRoute 
            path="/" 
            component={DashboardRoutes}
            isAuthenticated={user.logged}
            /> 
           
          </Switch>
        </div>
      </Router>
    )
}
