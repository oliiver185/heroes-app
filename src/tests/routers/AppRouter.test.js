import { mount, shallow } from "enzyme";
import { AuthContext } from "../../auth/AuthContext";
import { AppRouter } from "../../routers/AppRouter";

describe('Tests in <AppRouter/>', () => {
    
    
    const contextValue ={
        dispatch: jest.fn(),
        user: {
            logged: false
        }
    }
    test('should show login if is not athenticated', () => {
        


        const wrapper = mount(
            
            <AuthContext.Provider value={contextValue}>

                <AppRouter/>
            </AuthContext.Provider>
        );


        expect(wrapper).toMatchSnapshot();

        console.log(wrapper.html());

    });
    

    test('should show marvel component if is authenticated', () => {
        
        const contextValue ={
            dispatch: jest.fn(),
            user: {
                logged: true,
                name: 'oli'
            }
        }

        const wrapper = mount(
            
            <AuthContext.Provider value={contextValue}>

                <AppRouter/>
            </AuthContext.Provider>
        );


        // expect(wrapper).toMatchSnapshot();

        // console.log(wrapper.html());

        expect(wrapper.find('.navbar').exists()).toBe(true);

    });
    


});
