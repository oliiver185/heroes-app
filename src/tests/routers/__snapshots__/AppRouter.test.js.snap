// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <AppRouter/> should show login if is not athenticated 1`] = `
<div>
  <div
    className="container mt-5"
  >
    <h1>
      Login Screen
    </h1>
    <hr />
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Login
    </button>
  </div>
</div>
`;
